NLL: error from URL crate

In https://github.com/rust-lang/rust/issues/47596, @SimonSapin reported several NLL errors in dependencies of the servo crate. @Aaron1011 later minimized one of those errors into [this example](https://play.rust-lang.org/?gist=0bd31f0d318c0a4889f720fe54d7979d&version=undefined):

```rust
#![feature(nll)]

struct MyStruct<'a> {
    field: &'a mut (),
    field2: WithDrop
}

struct WithDrop;

impl Drop for WithDrop {

    fn drop(&mut self) {}

}

impl<'a> MyStruct<'a> {

    fn consume(self) -> &'a mut () { self.field }
}

fn main() {}
```

which yields:

```
error[E0597]: `*self.field` does not live long enough
  --> src/main.rs:18:38
   |
18 |     fn consume(self) -> &'a mut () { self.field }
   |                                      ^^^^^^^^^^  - borrowed value only lives until here
   |                                      |
   |                                      borrowed value does not live long enough
   |
   = note: borrowed value must be valid for lifetime '_#3r...
```

Removing feature(nll) makes the code work, so *probably* this is an NLL bug.

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -2073,7 +2073,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     /// currently in, when such distinction matters.
     fn each_borrow_involving_path<F>(
         &mut self,
-        _context: Context,
+        context: Context,
         access_place: (ShallowOrDeep, &Place<'tcx>),
         flow_state: &Flows<'cx, 'gcx, 'tcx>,
         mut op: F,
@@ -2085,20 +2085,50 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         // FIXME: analogous code in check_loans first maps `place` to
         // its base_path.
 
+        // When this function is called as a result of an `access_terminator` call attempting
+        // to drop a struct, if that struct does not have a destructor, then we need to check
+        // each of the fields in the struct. See #47703.
+        let (access, places) = if let ContextKind::Drop = context.kind {
+            let ty = place.ty(self.mir, self.tcx).to_ty(self.tcx);
+
+            match ty.sty {
+                ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
+                    let mut places = Vec::new();
+
+                    for (index, field) in def.all_fields().enumerate() {
+                        let proj = Projection {
+                            base: place.clone(),
+                            elem: ProjectionElem::Field(Field::new(index),
+                                                        field.ty(self.tcx, substs)),
+                        };
+
+                        places.push(Place::Projection(Box::new(proj)));
+                    }
+
+                    (ShallowOrDeep::Shallow(None), places)
+                },
+                _ => (access, vec![ place.clone() ]),
+            }
+        } else {
+            (access, vec![ place.clone() ])
+        };
+
         let data = flow_state.borrows.operator().borrows();
 
         // check for loan restricting path P being used. Accounts for
         // borrows of P, P.a.b, etc.
-        let mut elems_incoming = flow_state.borrows.elems_incoming();
-        while let Some(i) = elems_incoming.next() {
-            let borrowed = &data[i.borrow_index()];
+        for place in places {
+            let mut elems_incoming = flow_state.borrows.elems_incoming();
+            while let Some(i) = elems_incoming.next() {
+                let borrowed = &data[i.borrow_index()];
 
-            if self.places_conflict(&borrowed.borrowed_place, place, access) {
-                debug!("each_borrow_involving_path: {:?} @ {:?} vs. {:?}/{:?}",
-                       i, borrowed, place, access);
-                let ctrl = op(self, i, borrowed);
-                if ctrl == Control::Break {
-                    return;
+                if self.places_conflict(&borrowed.borrowed_place, &place, access) {
+                    debug!("each_borrow_involving_path: {:?} @ {:?} vs. {:?}/{:?}",
+                           i, borrowed, place, access);
+                    let ctrl = op(self, i, borrowed);
+                    if ctrl == Control::Break {
+                        return;
+                    }
                 }
             }
         }

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -711,6 +711,12 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
              self.tcx.sess.opts.debugging_opts.two_phase_beyond_autoref)
     }
 
+    /// Invokes `access_place` as appropriate for dropping the value
+    /// at `drop_place`. Note that the *actual* `Drop` in the MIR is
+    /// always for a variable (e.g., `Drop(x)`) -- but we recursively
+    /// break this variable down into subpaths (e.g., `Drop(x.foo)`)
+    /// to indicate more precisely which fields might actually be
+    /// accessed by a destructor.
     fn visit_terminator_drop(
         &mut self,
         loc: Location,
@@ -721,15 +727,16 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     ) {
         let ty = drop_place.ty(self.mir, self.tcx).to_ty(self.tcx);
         match ty.sty {
-            // When a struct is being dropped, we need to check whether it has a
-            // destructor, if it does, then we can call it, if it does not then we
-            // need to check the individual fields instead.
-            // See #47703.
+            // When a struct is being dropped, we need to check
+            // whether it has a destructor, if it does, then we can
+            // call it, if it does not then we need to check the
+            // individual fields instead. This way if `foo` has a
+            // destructor but `bar` does not, we will only check for
+            // borrows of `x.foo` and not `x.bar`. See #47703.
             ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
                 for (index, field) in def.all_fields().enumerate() {
                     let place = drop_place.clone();
                     let place = place.field(Field::new(index), field.ty(self.tcx, substs));
-                    let place = place.deref();
 
                     self.visit_terminator_drop(
                         loc,
@@ -741,13 +748,22 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 }
             },
             _ => {
-                self.access_place(
-                    ContextKind::Drop.new(loc),
-                    (drop_place, span),
-                    (Deep, Write(WriteKind::StorageDeadOrDrop)),
-                    LocalMutationIsAllowed::Yes,
-                    flow_state,
-                );
+                // We have now refined the type of the value being
+                // dropped (potentially) to just the type of a
+                // subfield; so check whether that field's type still
+                // "needs drop". If so, we assume that the destructor
+                // may access any data it likes (i.e., a Deep Write).
+                let gcx = self.tcx.global_tcx();
+                let erased_ty = gcx.lift(&self.tcx.erase_regions(&ty)).unwrap();
+                if erased_ty.needs_drop(gcx, self.param_env) {
+                    self.access_place(
+                        ContextKind::Drop.new(loc),
+                        (drop_place, span),
+                        (Deep, Write(WriteKind::StorageDeadOrDrop)),
+                        LocalMutationIsAllowed::Yes,
+                        flow_state,
+                    );
+                }
             },
         }
     }

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -727,35 +727,17 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             // See #47703.
             ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
                 for (index, field) in def.all_fields().enumerate() {
-                    let field_ty = field.ty(self.tcx, substs);
-                    let proj = Projection {
-                        base: drop_place.clone(),
-                        elem: ProjectionElem::Field(Field::new(index), field_ty),
-                    };
-                    let place = Place::Projection(Box::new(proj));
+                    let place = drop_place.clone();
+                    let place = place.field(Field::new(index), field.ty(self.tcx, substs));
+                    let place = place.deref();
 
-                    match field_ty.sty {
-                        // It may be the case that this issue occurs with a struct within a
-                        // struct, so we recurse to handle that.
-                        ty::TyAdt(def, _) if def.is_struct() && !def.has_dtor(self.tcx) => {
-                            self.visit_terminator_drop(
-                                loc,
-                                term,
-                                flow_state,
-                                &place,
-                                span,
-                            );
-                        },
-                        _ => {
-                            self.access_place(
-                                ContextKind::Drop.new(loc),
-                                (&place, span),
-                                (Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
-                                LocalMutationIsAllowed::Yes,
-                                flow_state,
-                            );
-                        },
-                    }
+                    self.visit_terminator_drop(
+                        loc,
+                        term,
+                        flow_state,
+                        &place,
+                        span,
+                    );
                 }
             },
             _ => {

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -463,13 +463,7 @@ impl<'cx, 'gcx, 'tcx> DataflowResultsConsumer<'cx, 'tcx> for MirBorrowckCtxt<'cx
                 target: _,
                 unwind: _,
             } => {
-                self.access_place(
-                    ContextKind::Drop.new(loc),
-                    (drop_place, span),
-                    (Deep, Write(WriteKind::StorageDeadOrDrop)),
-                    LocalMutationIsAllowed::Yes,
-                    flow_state,
-                );
+                self.visit_terminator_drop(loc, term, flow_state, drop_place, span);
             }
             TerminatorKind::DropAndReplace {
                 location: ref drop_place,
@@ -715,6 +709,65 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         self.tcx.sess.two_phase_borrows() &&
             (kind.allows_two_phase_borrow() ||
              self.tcx.sess.opts.debugging_opts.two_phase_beyond_autoref)
+    }
+
+    fn visit_terminator_drop(
+        &mut self,
+        loc: Location,
+        term: &Terminator<'tcx>,
+        flow_state: &Flows<'cx, 'gcx, 'tcx>,
+        drop_place: &Place<'tcx>,
+        span: Span,
+    ) {
+        let ty = drop_place.ty(self.mir, self.tcx).to_ty(self.tcx);
+        match ty.sty {
+            // When a struct is being dropped, we need to check whether it has a
+            // destructor, if it does, then we can call it, if it does not then we
+            // need to check the individual fields instead.
+            // See #47703.
+            ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
+                for (index, field) in def.all_fields().enumerate() {
+                    let field_ty = field.ty(self.tcx, substs);
+                    let proj = Projection {
+                        base: drop_place.clone(),
+                        elem: ProjectionElem::Field(Field::new(index), field_ty),
+                    };
+                    let place = Place::Projection(Box::new(proj));
+
+                    match field_ty.sty {
+                        // It may be the case that this issue occurs with a struct within a
+                        // struct, so we recurse to handle that.
+                        ty::TyAdt(def, _) if def.is_struct() && !def.has_dtor(self.tcx) => {
+                            self.visit_terminator_drop(
+                                loc,
+                                term,
+                                flow_state,
+                                &place,
+                                span,
+                            );
+                        },
+                        _ => {
+                            self.access_place(
+                                ContextKind::Drop.new(loc),
+                                (&place, span),
+                                (Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
+                                LocalMutationIsAllowed::Yes,
+                                flow_state,
+                            );
+                        },
+                    }
+                }
+            },
+            _ => {
+                self.access_place(
+                    ContextKind::Drop.new(loc),
+                    (drop_place, span),
+                    (Deep, Write(WriteKind::StorageDeadOrDrop)),
+                    LocalMutationIsAllowed::Yes,
+                    flow_state,
+                );
+            },
+        }
     }
 
     /// Checks an access to the given place to see if it is allowed. Examines the set of borrows
@@ -2073,7 +2126,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     /// currently in, when such distinction matters.
     fn each_borrow_involving_path<F>(
         &mut self,
-        context: Context,
+        _context: Context,
         access_place: (ShallowOrDeep, &Place<'tcx>),
         flow_state: &Flows<'cx, 'gcx, 'tcx>,
         mut op: F,
@@ -2085,50 +2138,20 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         // FIXME: analogous code in check_loans first maps `place` to
         // its base_path.
 
-        // When this function is called as a result of an `access_terminator` call attempting
-        // to drop a struct, if that struct does not have a destructor, then we need to check
-        // each of the fields in the struct. See #47703.
-        let (access, places) = if let ContextKind::Drop = context.kind {
-            let ty = place.ty(self.mir, self.tcx).to_ty(self.tcx);
-
-            match ty.sty {
-                ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
-                    let mut places = Vec::new();
-
-                    for (index, field) in def.all_fields().enumerate() {
-                        let proj = Projection {
-                            base: place.clone(),
-                            elem: ProjectionElem::Field(Field::new(index),
-                                                        field.ty(self.tcx, substs)),
-                        };
-
-                        places.push(Place::Projection(Box::new(proj)));
-                    }
-
-                    (ShallowOrDeep::Shallow(None), places)
-                },
-                _ => (access, vec![ place.clone() ]),
-            }
-        } else {
-            (access, vec![ place.clone() ])
-        };
-
         let data = flow_state.borrows.operator().borrows();
 
         // check for loan restricting path P being used. Accounts for
         // borrows of P, P.a.b, etc.
-        for place in places {
-            let mut elems_incoming = flow_state.borrows.elems_incoming();
-            while let Some(i) = elems_incoming.next() {
-                let borrowed = &data[i.borrow_index()];
+        let mut elems_incoming = flow_state.borrows.elems_incoming();
+        while let Some(i) = elems_incoming.next() {
+            let borrowed = &data[i.borrow_index()];
 
-                if self.places_conflict(&borrowed.borrowed_place, &place, access) {
-                    debug!("each_borrow_involving_path: {:?} @ {:?} vs. {:?}/{:?}",
-                           i, borrowed, place, access);
-                    let ctrl = op(self, i, borrowed);
-                    if ctrl == Control::Break {
-                        return;
-                    }
+            if self.places_conflict(&borrowed.borrowed_place, &place, access) {
+                debug!("each_borrow_involving_path: {:?} @ {:?} vs. {:?}/{:?}",
+                       i, borrowed, place, access);
+                let ctrl = op(self, i, borrowed);
+                if ctrl == Control::Break {
+                    return;
                 }
             }
         }

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -463,7 +463,20 @@ impl<'cx, 'gcx, 'tcx> DataflowResultsConsumer<'cx, 'tcx> for MirBorrowckCtxt<'cx
                 target: _,
                 unwind: _,
             } => {
-                self.visit_terminator_drop(loc, term, flow_state, drop_place, span);
+                let gcx = self.tcx.global_tcx();
+
+                // Compute the type with accurate region information.
+                let drop_place_ty = drop_place.ty(self.mir, self.tcx);
+
+                // Erase the regions.
+                let drop_place_ty = self.tcx.erase_regions(&drop_place_ty).to_ty(self.tcx);
+
+                // "Lift" into the gcx -- once regions are erased, this type should be in the
+                // global arenas; this "lift" operation basically just asserts that is true, but
+                // that is useful later.
+                let drop_place_ty = gcx.lift(&drop_place_ty).unwrap();
+
+                self.visit_terminator_drop(loc, term, flow_state, drop_place, drop_place_ty, span);
             }
             TerminatorKind::DropAndReplace {
                 location: ref drop_place,
@@ -723,10 +736,10 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         term: &Terminator<'tcx>,
         flow_state: &Flows<'cx, 'gcx, 'tcx>,
         drop_place: &Place<'tcx>,
+        erased_drop_place_ty: ty::Ty<'gcx>,
         span: Span,
     ) {
-        let ty = drop_place.ty(self.mir, self.tcx).to_ty(self.tcx);
-        match ty.sty {
+        match erased_drop_place_ty.sty {
             // When a struct is being dropped, we need to check
             // whether it has a destructor, if it does, then we can
             // call it, if it does not then we need to check the
@@ -735,14 +748,17 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             // borrows of `x.foo` and not `x.bar`. See #47703.
             ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
                 for (index, field) in def.all_fields().enumerate() {
-                    let place = drop_place.clone();
-                    let place = place.field(Field::new(index), field.ty(self.tcx, substs));
+                    let gcx = self.tcx.global_tcx();
+                    let field_ty = field.ty(gcx, substs);
+                    let field_ty = gcx.normalize_associated_type_in_env(&field_ty, self.param_env);
+                    let place = drop_place.clone().field(Field::new(index), field_ty);
 
                     self.visit_terminator_drop(
                         loc,
                         term,
                         flow_state,
                         &place,
+                        field_ty,
                         span,
                     );
                 }
@@ -754,8 +770,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 // "needs drop". If so, we assume that the destructor
                 // may access any data it likes (i.e., a Deep Write).
                 let gcx = self.tcx.global_tcx();
-                let erased_ty = gcx.lift(&self.tcx.erase_regions(&ty)).unwrap();
-                if erased_ty.needs_drop(gcx, self.param_env) {
+                if erased_drop_place_ty.needs_drop(gcx, self.param_env) {
                     self.access_place(
                         ContextKind::Drop.new(loc),
                         (drop_place, span),
@@ -2144,7 +2159,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         while let Some(i) = elems_incoming.next() {
             let borrowed = &data[i.borrow_index()];
 
-            if self.places_conflict(&borrowed.borrowed_place, &place, access) {
+            if self.places_conflict(&borrowed.borrowed_place, place, access) {
                 debug!("each_borrow_involving_path: {:?} @ {:?} vs. {:?}/{:?}",
                        i, borrowed, place, access);
                 let ctrl = op(self, i, borrowed);

--- a/src/test/run-pass/issue-47703-1.rs
+++ b/src/test/run-pass/issue-47703-1.rs
@@ -1,0 +1,33 @@
+// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+
+struct AtomicRefMut<'a> {
+    value: &'a mut i32,
+    borrow: AtomicBorrowRefMut,
+}
+
+struct AtomicBorrowRefMut {
+}
+
+impl Drop for AtomicBorrowRefMut {
+    fn drop(&mut self) {
+    }
+}
+
+fn map(orig: AtomicRefMut) -> AtomicRefMut {
+    AtomicRefMut {
+        value: orig.value,
+        borrow: orig.borrow,
+    }
+}
+
+fn main() {}

--- a/src/test/run-pass/issue-47703.rs
+++ b/src/test/run-pass/issue-47703.rs
@@ -1,0 +1,28 @@
+// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+
+struct MyStruct<'a> {
+    field: &'a mut (),
+    field2: WithDrop
+}
+
+struct WithDrop;
+
+impl Drop for WithDrop {
+    fn drop(&mut self) {}
+}
+
+impl<'a> MyStruct<'a> {
+    fn consume(self) -> &'a mut () { self.field }
+}
+
+fn main() {}

